### PR TITLE
Dockerize

### PR DESCRIPTION
**Pull request by bmhenry**
I saw issue #26 and had similar desires to @danieleades. This pull request needs some review and is likely a bit of a work in  progress, as it needs some input as to default options, etc.

This adds a `docker-compose`-compatible setup to run Alexandrie in a Docker container. It does so in a way that makes it relatively easy for the user to manage the location of the crate storage, database (assuming sqlite/a database file), crate index, and config files.

The biggest thing this PR lacks is database management. It works for a SQLite database, because in that case the database is a single file and it's easy to tell Alexandrie to put that file in a certain location. I also have no experience with `diesel`, so I am guessing that some extra steps need to be added to handle automatic migrations in the Docker container.

If you've already made more progress into a containerization than this, feel free to discard this. Otherwise, hopefully this is at least a good first step!

### CODE REVIEW

**Comment by danieleades** [comment]
Fantastic! I never found time to look at this beyond an initial investigation.
I did wonder if it belongs in a separate repo, because I think it has the potential to grow arms and legs.
I guess that's a question for the maintainers

**Comment by bmhenry** [comment]
It may very well end up needing to be its own repo. There's probably some benefit to keeping it all in one if possible, though. For example, a GitHub Action could be added that tries to build and run the docker container before a PR can be merged, to test and see if there are any breaking API changes or anything. That said, the build does take a while, so it might not fit in GitHub's restrictions.

**Comment by Hirevo** [comment]
As for the question of whether we should move all this into a separate repository:
I agree that keeping in the same repository makes CI easier.  
Currently, the CI doesn't tries to build to project because Github Actions runner images don't include SQLite or PostgreSQL at the moment.  
So, I think, with this PR, we may start to do these automated builds (to make sure that changes to the schema or additions to migrations works on all database vendors).  
Maybe the image configurations could be organized like the current `migrations/` folder, like:
```
docker/
|-- mysql/
|   '-- ... (configuration files for images with MySQL)
|-- postgres/
|   '-- ... (configuration files for images with PostgreSQL)
'-- sqlite/
    '-- ... (configuration files for images with SQLite)
```
The repository currently weighs around 4 MB, so it is not too big yet and if all we add here is configuration files, I don't think it is too much of an issue.  
Although, I think I'll add READMEs in the different subdirectories to ease navigation and documenting more for newcomers.

File: alexandrie.toml [hunk 1]
```diff
@@ -1,29 +1,29 @@
...
 
 [index]
 type = "command-line"
-path = "crate-index"
+path = "appdata/crate-index"
 
 [storage]
 type = "disk"
-path = "crate-storage"
+path = "appdata/crate-storage"
```
[the remark this answers is not in the document]
Comment (bmhenry) [reply alexandrie.toml]: Whoops, forgot to revert this file. Yeah, you're right.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,71 @@
+#
+# Dockerfile for the Alexandrie crate registry application
+#
+# The output docker image will assume the default Cargo.toml options
+# (i.e., sqlite3 database)
+#
+
+### First stage: build the application
+FROM rust as builder
```
[the remark this answers is not in the document]
Comment (bmhenry) [reply Dockerfile]: Yeah, that's a good idea. It's another thing to remember to update, but probably worth it.

**Comment by bmhenry** [comment]
Currently, I'm mounting the `diesel.toml` config into the container, but as I look into diesel a little it seems like that may be completely unnecessary? Is that file only used by you when you're generating migrations, and never by the app at runtime?

**Comment by Hirevo** [comment]
Yes, `diesel.toml` is only used by the `diesel-cli` tool, that I use to generate migrations or manually replay them.
It is not read by the application in any way, so you're correct, we can indeed omit it from the Docker image.

**Comment by bmhenry** [comment]
Posting an update to make sure people know I haven't abandoned this:

* I'm working on separate mysql and postgres docker configurations, using the existing Docker containers on docker hub rather than running the databases inside the alexandrie container. I'm running into issues with mysql complaining that the `alexandrie` database can't be found
* I'm also working on kubectl yaml files so this can be ran on a Kubernetes cluster. Hopefully I'll be able to add some scalability/availability in so that in the future there's the possibility for no-/low-downtime upgrades

**Comment by bmhenry** [comment]
@Hirevo I think I'm mostly there, but there is a hitch with database creation. When you locally have ran with mysql or postgres, do you do a `diesel database setup` to create the database the first time? It doesn't appear that Alexandrie will create the `alexandrie` database if there isn't one already. 

It looks like `diesel_cli` does that [here](https://github.com/diesel-rs/diesel/blob/master/diesel_cli/src/database.rs#L153), would it be possible for Alexandrie to do something pretty similar using the URL it gets via the config? I can understand if that's not something you think Alexandrie should do, but it would be useful for a nothing-to-running workflow.

**Comment by Hirevo** [comment]
Yes indeed, we don't do the database creation part.  
I always do an initial `CREATE DATABASE alexandrie;` in the database REPL.  
Diesel itself doesn't have any database creation query built-in but you're right, we could steal the impls from diesel_cli if needed.  
If we were to do this, I think it would be with an opt-in configuration option, because if you made a typo in the connection url, automatically creating it would silence an error you may have expected.

**Comment by bmhenry** [comment]
Thanks to #33, I now have Docker working for all three databases! There's a handy script that wraps the `docker-compose` calls under the hood, as there are multiple compose files depending on the database a user wants. All three store use volumes to persist data between runs, so you can stop and restart Alexandrie without issue.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    if [ "${DATABASE}" = "sqlite" ]; then apt install -y sqlite3; fi && \
+    if [ "${DATABASE}" = "postgresql" ]; then apt install -y  postgresql; fi && \
+    if [ "${DATABASE}" = "mysql" ]; then apt install -y default-mysql-server default-mysql-client; fi && \
+    apt-get clean && rm -rf /var/lib/apt/lists/
+
+
+# copy run files
+COPY --from=builder /alexandrie/target/release/alexandrie /usr/bin/alexandrie
+# copy docker_cli
+COPY --from=builder /usr/local/cargo/bin/diesel /usr/bin/diesel
```
[the remark this answers is not in the document]
Comment (Hirevo) [reply Dockerfile]: I just noticed that you have the following code for the SQLite case:
```
if [ "${DATABASE}" = "sqlite" ]; then \
    apt install -y sqlite3 libsqlite3-dev; \
    mkdir -p /usr/local/cargo/bin/; \
    touch /usr/local/cargo/bin; \
fi
```
Maybe the final `touch` command should have been:
```
touch /usr/local/cargo/bin/diesel_cli;
```
to make the `COPY` succeed.

**Comment by bmhenry** [comment]
Whoops, I must've forgotten to retest sqlite. Good catches, I'll fix these and push the commit

File: run_docker.sh [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    exit 1
+fi
+
+# run docker-compose
+if [ "$CMD" = "up" ]; then
+    if [ "$DO_CLEAN" = "true" ]; then
+        DATABASE=$DATABASE docker-compose ${FILES} build --no-cache
+        DATABASE=$DATABASE docker-compose ${FILES} up ${DAEMON}
+    else
+        DATABASE=$DATABASE docker-compose ${FILES} up ${DO_BUILD}
```
[the remark this answers is not in the document]
Comment (bmhenry) [reply run_docker.sh]: Good catch

File: Dockerfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (bmhenry) [reply Dockerfile]: Whoops, yep.

**Comment by Hirevo** [comment]
Oops, one thing Oops, one thing I remembered wondering about but forgot to mention in the review is that starting an image build can be slowed down by the transfer of the entire build context (the contents of the current directory) to the Docker daemon because of the large size of the `target/` folder (which can sit at around 4 Go, at times).  
So I thought we could maybe use a `.dockerignore`, because I would suspect that the `target/` folder is not reusable across machines (even with the same processor architecture and executable formats, because the different paths/availability of some dynamic libraries could make reusing intermediate build artifacts challenging).

**Comment by bmhenry** [comment]
That seems like a great idea to me, I can put in a PR for that. I hadn't though of limiting the context.  `target` is definitely large, and we could probably skip the small stuff from the `.gitignore` as well

**Pull request merged**